### DISCUSSION TITLE
Adding Mesh to simulation

### DISCUSSION THREAD

**Discussion by Fax3D** [bepu/bepuphysics2#196] [Q&A]
Hi, 
Could you please show me how to add a mesh to the simulation?

Here my code snippet where i extract vertices from my model to create a mesh... and then i dont' know how to use this mesh to add a new "body" to the simulation...

` 
physics.BufferPool.Take<Triangle>(_trianglesCount, out var _triangleBuffer);
  for (int i = 0; i < _trianglesCount; i++)
  {
      ref Triangle t = ref _triangleBuffer[i];

      t.A.X = _vertices[_indices[i * 3 + 0]].Position.X;
      t.A.Y = _vertices[_indices[i * 3 + 0]].Position.Y;
      t.A.Z = _vertices[_indices[i * 3 + 0]].Position.Z;

      t.B.X = _vertices[_indices[i * 3 + 1]].Position.X;
      t.B.Y = _vertices[_indices[i * 3 + 1]].Position.Y;
      t.B.Z = _vertices[_indices[i * 3 + 1]].Position.Z;

      t.C.X = _vertices[_indices[i * 3 + 2]].Position.X;
      t.C.Y = _vertices[_indices[i * 3 + 2]].Position.Y;
      t.C.Z = _vertices[_indices[i * 3 + 2]].Position.Z;
  }
  var _mesh = new Mesh(_triangleBuffer.Slice(0, _trianglesCount), System.Numerics.Vector3.One, physics.BufferPool);
  physics.Bodies.Add(?)
`

Thanks for your help!

**Comment by RossNordby** [comment]
A `Mesh` is a shape, like a `Sphere` or `Box`. You'll need to add it first:
`var shapeIndex = Simulation.Shapes.Add(mesh);`
Then, when creating a body, provide the shapeIndex to the `CollidableDescription` in the `BodyDescription` or `StaticDescription`.

So if creating a body:
```cs
Simulation.Bodies.Add(BodyDescription.CreateDynamic(
    new RigidPose(new Vector3(0, -15, 0), QuaternionEx.CreateFromAxisAngle(new Vector3(0, 1, 0), MathF.PI / 2)), 
    mesh.ComputeClosedInertia(1), 
    new CollidableDescription(Simulation.Shapes.Add(mesh)), 
    new BodyActivityDescription(0.01f)));
```
Or a static:
```cs
Simulation.Statics.Add(new StaticDescription(
    new Vector3(0, -15, 0), QuaternionEx.CreateFromAxisAngle(new Vector3(0, 1, 0), MathF.PI / 2),
    new CollidableDescription(Simulation.Shapes.Add(mesh))));
```

2.4 also has some convenience casts that can make the description of bodies shorter. For example, you can directly provide the shape's `TypedIndex` as a `CollidableDescription` and a `float` for the `BodyActivityDescription`:
```cs
Simulation.Bodies.Add(BodyDescription.CreateDynamic(
    new RigidPose(new Vector3(0, -15, 0), QuaternionEx.CreateFromAxisAngle(new Vector3(0, 1, 0), MathF.PI / 2)),
    mesh.ComputeClosedInertia(1), Simulation.Shapes.Add(mesh), 0.01f));
```
```cs
Simulation.Statics.Add(new StaticDescription(
    new Vector3(0, -15, 0), QuaternionEx.CreateFromAxisAngle(new Vector3(0, 1, 0), MathF.PI / 2), 
    Simulation.Shapes.Add(mesh)));
```

You can find more examples in the demos. The idea is that shapes can be shared across many bodies/statics.

Also, I generally recommend against using `Mesh` with dynamic bodies. See the bit about meshes in [the performance tips.](https://github.com/bepu/bepuphysics2/blob/master/Documentation/PerformanceTips.md) Approximating meshes with simpler shapes, especially for dynamics, is a big win.
